sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/BufferHolder.java

```diff
@@ -55,6 +58,15 @@ public BufferHolder(UnsafeRow row, int initialSize) {
     this.buffer = new byte[fixedSize + initialSize];
     this.row = row;
     this.row.pointTo(buffer, buffer.length);
+    this.array = null;
+  }
+
+  public BufferHolder(UnsafeArrayData array, long numElements) {
```

Comment:
@kiszk We only need to use `BufferHolder` to allocate a big enough buffer. Then we use `UnsafeArrayWriter` to write primitive elements into the buffer. After that, we create a `UnsafeArrayData` and point it to the buffer as its backing data.

So we don't need to add such change into `BufferHolder` like this. Just a new constructor is needed.